test/functional/forall/test-forall-rangesegment.hpp

```diff
@@ -0,0 +1,73 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef __TEST_FORALL_RANGESEGMENT_HPP__
+#define __TEST_FORALL_RANGESEGMENT_HPP__
+
+#include "RAJA/RAJA.hpp"
+#include "camp/resource.hpp"
+
+template<typename T>
+void ForallRangeSegmentFunctionalTest_host(T first, T last)
+{
+  RAJA::TypedRangeSegment<T> r1(first, last);
+  RAJA::Index_type N = r1.end() - r1.begin();
+
+  camp::resources::Resource res{camp::resources::Host()};
```

Comment:
Rather than having _host and _cuda variants, one solution I discussed with @rchen20 is always having two resource objects. Once for the host and the other for a device, here a device can be again the host, cuda or hip. You can then pass in the device resource manager as an argument in the functional test function. Inside the actual function we would still go through the motions of copying to and from the device. Does this make sense?

Reply:
I think you always want a host copy for checking the data, then you allocate a "working" copy based on the Platform/Policy - you copy the data back to the "check" array for doing the ASSERT.

The Umpire tests are structured in a similar way.

Reply:
I agree with @artv3, especially in the header files that define the tests. I think our goal should be to have no programming model or architecture specific code in those files and use the source files to specialize those tests for the different back-end cases.

Reply:
Okay thanks all, I'll have another go at this.